_posts/2019-03-08-code-quality-inspection-for-clojure-with-sonarqube.md

```diff
@@ -0,0 +1,194 @@
+
+## Clojure code quality tools
+
+### Eastwood
+[Eastwood](https://github.com/jonase/eastwood) is a lint tool for Clojure (no CLJS support) which detects many kinds of potentially problematic code. 
+
+For example next Clojure code contains a potential bug:
+
+```clojure
+(if (and x)
+  (foo))
+```
+
+When running ```lein eastwood``` it produces next result:
+```
+== Eastwood 0.3.3 Clojure 1.8.0 JVM 11.0.1
+Directories scanned for source files:
+ src test
+== Linting clojure-sonar-example.core ==
+Entering directory `/Users/heikkiha/src/sonarqube/clojure-sonar-example'
+src/clojure_sonar_example/core.clj:3:5: suspicious-expression: and called with 1 args.  (and x) always returns x.  Perhaps there are misplaced parentheses?
+```
+ 
+Eastwood may cause false positives. Fortunately it is possible to suppress linter rules which may cause any false positives.
+ 
+### Kibit
+[Kibit](https://github.com/jonase/eastwood) is a static code analyzer which detects code that could be rewritten with a more idiomatic function or macro. It supports also ClojureScript. 
+
+Example: 
+
+```clojure
+(> x 0) 
+```
+
+```lein kibit``` produces next result:
+
+```
+At src/clojure_sonar_example/core.clj:13:
+   Consider using:
+     (pos? x)
+   instead of:
+     (> x 0)
+```
+
+I think Kibit is most useful for beginning Clojure programmers but it contains some good advices for more experienced developers also. 
+
+### Cloverage
+
+[Cloverage](https://github.com/cloverage/cloverage) is a code coverage tool for Clojure which runs the tests ot a program 
+and calculates line and form coverage for namespaces. 
+
+Sample output looks like this:
+
+|                    Namespace | % Forms | % Lines |
+|------------------------------|---------|---------|
+|foo.bar                       |   70.00 |   84.62 |
+|foo.baz                       |   94.61 |   96.89 |
+|foo.config                    |   60.71 |   80.65 |
+|foo.router                    |   31.75 |   69.62 |
+|foo.login                     |   68.67 |   82.76 |
+|foo.util                      |    2.89 |   20.00 |
+|------------------------------|---------|---------|
+|                    ALL FILES |   67.36 |   82.93 |
+
+
+### Lein-nvd
+
+[Lein-nvd](https://github.com/rm-hull/lein-nvd) is a dependency-checker plugin whichs checks JARS in the programs classpath for known vulnerabilites against 
+the [National Vulnerability Database](https://nvd.nist.gov/). 
+
+Output of Lein-nvd gives status of the JARs with found vulnerabilities:
+
+
+| dependency             | status|
+|------------------------|-------|
+| bcpkix-jdk15on-1.58.jar | OK    | 
+| bcprov-jdk15on-1.58.jar| CVE-2017-13098, CVE-2018-1000180, CVE-2018-1000613|
+
+It should be noted that Lein-nvd does not exactly tell which dependency or its transitive dependency pulls a vulnerable JAR file. Most time JAR name matches a dependency. The full dependency tree can be inspected by running ```lein deps :tree``` command.
```

Comment:
"does not tell exactly which"
"direct or transitive dependency"
"running *the* `lein deps :tree` command"